### ISSUE TITLE
Precompilation error on Julia v0.6

### ISSUE THREAD

**Issue by juliohm** [JuliaStats/Distributions.jl#631]
Currently I am unable to compile Distributions.jl on Julia v0.6:

```julia
julia> using Distributions
INFO: Recompiling stale cache file /home/juliohm/.julia/lib/v0.6/Distributions.ji for module Distributions.
WARNING: The call to compilecache failed to create a usable precompiled cache file for module Distributions. Got:
WARNING: Module FFTW uuid did not match cache file.
ERROR: LoadError: Declaring __precompile__(true) is only allowed in module files being imported.
Stacktrace:
 [1] __precompile__(::Bool) at ./loading.jl:335
 [2] include_from_node1(::String) at ./loading.jl:569
 [3] eval(::Module, ::Any) at ./boot.jl:235
 [4] _require(::Symbol) at ./loading.jl:483
 [5] require(::Symbol) at ./loading.jl:398
while loading /home/juliohm/.julia/v0.6/Distributions/src/Distributions.jl, in expression starting on line 1
```

Similar issue has happened with `Combinatorics.jl`: https://github.com/JuliaMath/Combinatorics.jl/issues/48

What exactly is happening? How can I fix this myself?

**Comment by ararslan** [comment]
You might have to delete the contents of ~/.julia/lib/v0.6.

**Comment by juliohm** [comment]
Thank you @ararslan , I deleted it, but the problem remains. Any other suggestion that I can try locally?

**Comment by ararslan** [comment]
Hm, I'm not sure, I can't reproduce the failure locally. What's your `Pkg.status()` and `versioninfo()`?

**Comment by juliohm** [comment]
This is the output of both commands:

```julia
julia> versioninfo()
Julia Version 0.6.1-pre.0
Commit dcf39a1dda* (2017-06-19 13:06 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-6500U CPU @ 2.50GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, skylake)
julia> Pkg.status("Distributions")
 - Distributions                 0.14.0             master
julia> Pkg.status("FFTW")
 - FFTW                          0.0.2
```

And the version of fftw on my system in case it is relevant:

```
$ pacman -Qi fftw
Name            : fftw
Version         : 3.3.6-2
Description     : A library for computing the discrete Fourier transform (DFT)
Architecture    : x86_64
URL             : http://www.fftw.org/
Licenses        : GPL2
Groups          : None
Provides        : None
Depends On      : bash  gcc-libs
Optional Deps   : None
Required By     : alsa-utils  libofa
Optional For    : None
Conflicts With  : None
Replaces        : None
Installed Size  : 5.54 MiB
Packager        : Antonio Rojas <arojas@archlinux.org>
Build Date      : Sun 02 Apr 2017 02:16:46 PM PDT
Install Date    : Mon 03 Apr 2017 10:49:03 AM PDT
Install Reason  : Installed as a dependency for another package
Install Script  : No
Validated By    : Signature
```

**Comment by ararslan** [comment]
Can you try it from the 0.6 release binary rather than a build of the release-0.6 branch?

**Comment by juliohm** [comment]
Same issue on the precompiled binary, just tried it.

What exactly is `.julia/lib` and `.julia/.cache`? Can I safely delete the latter and try again?

**Comment by ararslan** [comment]
Uhhhhh sure, give it a try ¯\\\_(ツ)\_/¯

**Comment by ararslan** [comment]
lib is the precompiled cache files (.ji) and .cache is the git tracking info for each installed package. So maybe don't delete the latter...

**Comment by juliohm** [comment]
Tried it, nothing changed. And apparently this didn't cause any issue locally, I am able to load other packages and run the tests without a `.cache` folder.

**Comment by juliohm** [comment]
A `Pkg.update()` will create the `.cache` again, so I am back to the original state, any other place where you think the issue lies?

**Comment by nalimilan** [comment]
Can you try `using FFTW` in a clean session?

**Comment by juliohm** [comment]
Thank you @nalimilan, just tried it, and everything works. I get an expected warning:

```julia
julia> using FFTW
WARNING: imported binding for FFTW overwritten in module Main
```

**Comment by juliohm** [comment]
In the other issue I pointed (https://github.com/JuliaMath/Combinatorics.jl/issues/48), they solved it by replacing `Iterators.jl` with `IterTools.jl`. Is it an issue with `FFTW.jl` or the fact that it is now part of Base?

**Comment by ararslan** [comment]
Yes it's likely because FFTW.jl has the same name as the FFTW submodule in Base.

**Comment by juliohm** [comment]
I don't understand why this issue is only affecting me though, are you guys using Julia v0.6 as well?

**Comment by ararslan** [comment]
I am. I don't think anyone is quite sure what exactly causes this issue, though I think some people have had luck nuking their entire ~/.julia folder. Quite unfortunate but a possible solution.

**Comment by nalimilan** [comment]
Just a guess, but maybe the failure happens when some packages which have been precompiled and are dependencies of Distributions use functions from `Base.FFTW`? Then it would depend on the order in which packages are precompiled, which itself depends on the packages you have installed and the order in which you used them.

**Comment by juliohm** [comment]
Just tried deleting the entire `.julia` folder, the issue is still there. I appreciate if someone can work on this issue with FFTW, my work is stuck on this front.

**Comment by juliohm** [comment]
Since FFTW is part of Base now, shouldn't Distributions.jl use it instead of the FFTW.jl package?

**Comment by KristofferC** [comment]
FFTW is no longer part of Base.

**Comment by ararslan** [comment]
Yeah, it's the opposite situation: FFTW was part of Base up until quite recently, in 0.7-DEV. To support Julia nightly we need to add a dependency on the FFTW package, which contains the code moved out of Base.

**Comment by ChrisRackauckas** [comment]
With the Combinatorics/Iterators issue, `using Iterators, Combinatorics` before `using ...` fixed it. So for now, `using FFTW`, `using Distributions` is probably the quick workaround.

But it's fundamentally an issue with naming that hopefully will get solved. See https://github.com/JuliaLang/julia/issues/21969

**Comment by juliohm** [comment]
This is already of big help @ChrisRackauckas, thank you for sharing. I can proceed with my work meanwhile.

**Comment by ChrisRackauckas** [comment]
Can this change be reverted until FFTW.jl stabilizes a bit and gets some of its kinks worked out? This error is quite annoying to have in a pretty widely used package, and it doesn't build on RHEL/Fedora/CentOS. Right now the current release requires FFTW and does get in the way of normal v0.6 usage only for the benefit of users on the nightly. I think it should have a release version which is solid on v0.6 and a v0.7 branch for users to checkout if needed, at least until this is all sorted out.

**Comment by ararslan** [comment]
Which change do you mean, the deprecation in Base or the addition of FFTW as a dependency for Distributions? https://github.com/JuliaMath/FFTW.jl/issues/17 would get this working, there's just significant disagreement over the best approach to take on that front.

I had also suggested simply using a pure Julia implementation of the N<sup>2</sup> algorithm to avoid a dependency on FFTW altogether, since the only use of an FFT in this package is in computing the Poisson binomial PDF.

**Comment by ChrisRackauckas** [comment]
I was thinking that you should just create a new release which doesn't have FFTW.jl which won't work on v0.6, and have a branch with FFTW.jl setup and waiting for the solution to be worked out before merging and releasing with it. I think the majority of people are on v0.6 so this'll help a lot of non-power users who don't know the workaround, and in the meantime it's not a difficult branch to maintain.

**Comment by mauro3** [comment]
From scanning this long thread: One work-around for this is to delete lines 
https://github.com/JuliaStats/Distributions.jl/blob/0f89af2aa90de4d6e0341aad54279ce2a9f5efb7/REQUIRE#L9
and
https://github.com/JuliaStats/Distributions.jl/blob/0f89af2aa90de4d6e0341aad54279ce2a9f5efb7/src/Distributions.jl#L11

Is this the preferred work-around?  Maybe the preferred work-around could be inserted into the top post (@juliohm).

**Comment by ararslan** [comment]
Deleting a dependency that's present in the package isn't really a good workaround...

**Comment by mauro3** [comment]
Yep, that is why I'm asking for a good work around.  Note though it passes the tests fine.  If that should not be the case, then a test should be added.

**Comment by liamhenry** [comment]
Julia Version 0.6.0
Commit 903644385b* (2017-06-19 13:05 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i5-2540M CPU @ 2.60GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Sandybridge)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, sandybridge)


Pkg.status("Distributions")

 - Distributions                 0.14.2


Was experiencing similar problems. I am not as savvy to mess about with manually deleting .julia libraries or anything, but I followed the discussions mentioned here and couldn't resolve it. 

The solution I got was inspired by another issue, https://github.com/JuliaLang/julia/issues/19017 .
I turned it off and on again... and it now works!

From messing about, I don't remember the original error message, but the one I was getting most recently was along the lines of: 
"WARNING: Module Compat with uuid 9269499719553 is missing from the cache"
